File: packages/google_mlkit_object_detection/ios/Classes/GoogleMlKitObjectDetectionPlugin.m
#import "GoogleMlKitObjectDetectionPlugin.h"
#import <MLKitCommon/MLKitCommon.h>
#import <MLKitObjectDetection/MLKitObjectDetection.h>
#import <MLKitObjectDetectionCommon/MLKitObjectDetectionCommon.h>
#import <MLKitObjectDetectionCustom/MLKitObjectDetectionCustom.h>
#import <google_mlkit_commons/GoogleMlKitCommonsPlugin.h>

#if MLKIT_FIREBASE_MODELS
#import <MLKitLinkFirebase/MLKitLinkFirebase.h>
#endif

#define channelName @"google_mlkit_object_detector"
#define startObjectDetector @"vision#startObjectDetector"
#define closeObjectDetector @"vision#closeObjectDetector"
#define manageFirebaseModels @"vision#manageFirebaseModels"

@implementation GoogleMlKitObjectDetectionPlugin {
    NSMutableDictionary *instances;
    GenericModelManager *genericModelManager;
}

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel
                                     methodChannelWithName:channelName
                                     binaryMessenger:[registrar messenger]];
    GoogleMlKitObjectDetectionPlugin* instance = [[GoogleMlKitObjectDetectionPlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (id)init {
    self = [super init];
    if (self)
        instances = [NSMutableDictionary dictionary];
    return  self;
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    if ([call.method isEqualToString:startObjectDetector]) {
        [self handleDetection:call result:result];
    } else if ([call.method isEqualToString:closeObjectDetector]) {
        NSString *uid = call.arguments[@"id"];
        [instances removeObjectForKey:uid];
        result(NULL);
    } else if ([call.method isEqualToString:manageFirebaseModels]) {
#if MLKIT_FIREBASE_MODELS
        [self manageModel:call result:result];
#else
        result([FlutterError errorWithCode:@"ERROR_MISSING_MLKIT_FIREBASE_MODELS" message:@"You must define MLKIT_FIREBASE_MODELS=1 in your Podfile." details:nil]);
#endif
    } else {
        result(FlutterMethodNotImplemented);
    }
}

- (void)handleDetection:(FlutterMethodCall *)call result:(FlutterResult)result {
    MLKVisionImage *image = [MLKVisionImage visionImageFromData:call.arguments[@"imageData"]];
    
    NSString *uid = call.arguments[@"id"];
    MLKObjectDetector *objectDetector = [instances objectForKey:uid];
    if (objectDetector == NULL) {
        NSDictionary *dictionary = call.arguments[@"options"];
        NSString *type = dictionary[@"type"];
        if ([@"base" isEqualToString:type]) {
            MLKObjectDetectorOptions *options = [self getDefaultOptions:dictionary];
            objectDetector = [MLKObjectDetector objectDetectorWithOptions:options];
        } else if ([@"local" isEqualToString:type]) {
            MLKCustomObjectDetectorOptions *options = [self getLocalOptions:dictionary];
            objectDetector = [MLKObjectDetector objectDetectorWithOptions:options];
        } else if ([@"remote" isEqualToString:type]) {
#if MLKIT_FIREBASE_MODELS
            MLKCustomObjectDetectorOptions *options = [self getRemoteOptions:dictionary];
            if (options == NULL) {
                FlutterError *error = [FlutterError errorWithCode:@"Error Model has not been downloaded yet"
                                                          message:@"Model has not been downloaded yet"
                                                          details:@"Model has not been downloaded yet"];
                result(error);
                return;
            }
            objectDetector = [MLKObjectDetector objectDetectorWithOptions:options];
#else
            result([FlutterError errorWithCode:@"ERROR_MISSING_MLKIT_FIREBASE_MODELS" message:@"You must define MLKIT_FIREBASE_MODELS=1 in your Podfile." details:nil]);
#endif
        } else {
            NSString *error = [NSString stringWithFormat:@"Invalid model type: %@", type];
            result([FlutterError errorWithCode:type
                                       message:error
                                       details:error]);
            return;
        }
        instances[uid] = objectDetector;
    }
    
    [objectDetector processImage:image
                      completion:^(NSArray *_Nullable objects,
                                   NSError *_Nullable error) {
        if (error) {
            result(getFlutterError(error));
            return;
        } else if (!objects) {
            result(@[]);
            return;
        }
        
        NSMutableArray *objectsData = [NSMutableArray array];
        for (MLKObject *object in objects) {
            NSMutableArray *labels = [NSMutableArray array];
            for (MLKObjectLabel *label in object.labels) {
                [labels addObject:@{
                    @"index" : @(label.index),
                    @"text" : label.text,
                    @"confidence" : @(label.confidence),
                }];
            }
            NSMutableDictionary *data = [NSMutableDictionary dictionaryWithDictionary:@{
                @"rect" : @{
                    @"left" : @(object.frame.origin.x),
                    @"top" : @(object.frame.origin.y),
                    @"right" : @(object.frame.origin.x + object.frame.size.width),
                    @"bottom" : @(object.frame.origin.y + object.frame.size.height)
                },
                @"labels" : labels
            }];
            if (object.trackingID != NULL) {
                data[@"trackingId"] = object.trackingID;
            }
            [objectsData addObject:data];
        }
        
        result(objectsData);
    }];
}

- (MLKObjectDetectorOptions *)getDefaultOptions:(NSDictionary *)dictionary {
    NSNumber *mode = dictionary[@"mode"];
    BOOL classify = [[dictionary objectForKey:@"classify"] boolValue];
    BOOL multiple = [[dictionary objectForKey:@"multiple"] boolValue];
    
    MLKObjectDetectorOptions *options = [[MLKObjectDetectorOptions alloc] init];
    options.detectorMode = mode.intValue == 0 ? MLKObjectDetectorModeStream : MLKObjectDetectorModeSingleImage;
    options.shouldEnableClassification = classify;
    options.shouldEnableMultipleObjects = multiple;
    return options;
}

- (MLKCustomObjectDetectorOptions *)getLocalOptions:(NSDictionary *)dictionary {
    NSNumber *mode = dictionary[@"mode"];
    BOOL classify = [[dictionary objectForKey:@"classify"] boolValue];
    BOOL multiple = [[dictionary objectForKey:@"multiple"] boolValue];
    NSNumber *threshold = dictionary[@"threshold"];
    NSNumber *maxLabels = dictionary[@"maxLabels"];
    NSString *path = dictionary[@"path"];
    
    MLKLocalModel *localModel = [[MLKLocalModel alloc] initWithPath:path];
    MLKCustomObjectDetectorOptions *options =  [[MLKCustomObjectDetectorOptions alloc] initWithLocalModel:localModel];
    options.detectorMode = mode.intValue == 0 ? MLKObjectDetectorModeStream : MLKObjectDetectorModeSingleImage;
    options.shouldEnableClassification = classify;
    options.shouldEnableMultipleObjects = multiple;
    options.classificationConfidenceThreshold = threshold;
    options.maxPerObjectLabelCount = maxLabels.integerValue;
    return options;
}

#if MLKIT_FIREBASE_MODELS
- (MLKCustomObjectDetectorOptions *)getRemoteOptions:(NSDictionary *)dictionary {
    NSNumber *mode = dictionary[@"mode"];
    BOOL classify = [[dictionary objectForKey:@"classify"] boolValue];
    BOOL multiple = [[dictionary objectForKey:@"multiple"] boolValue];
    NSNumber *threshold = dictionary[@"threshold"];
    NSNumber *maxLabels = dictionary[@"maxLabels"];
    NSString *modelName = dictionary[@"modelName"];
    
    MLKFirebaseModelSource *firebaseModelSource = [[MLKFirebaseModelSource alloc] initWithName:modelName];
    MLKCustomRemoteModel *remoteModel = [[MLKCustomRemoteModel alloc] initWithRemoteModelSource:firebaseModelSource];
    
    MLKModelManager *modelManager = [MLKModelManager modelManager];
    BOOL isModelDownloaded = [modelManager isModelDownloaded:remoteModel];
    if (!isModelDownloaded) {
        return NULL;
    }
    
    MLKCustomObjectDetectorOptions *options = [[MLKCustomObjectDetectorOptions alloc] initWithRemoteModel:remoteModel];
    options.detectorMode = mode.intValue == 0 ? MLKObjectDetectorModeStream : MLKObjectDetectorModeSingleImage;
    options.shouldEnableClassification = classify;
    options.shouldEnableMultipleObjects = multiple;
    options.classificationConfidenceThreshold = threshold;
    options.maxPerObjectLabelCount = maxLabels.integerValue;
    return options;
}

- (void)manageModel:(FlutterMethodCall *)call result:(FlutterResult)result {
    NSString *modelTag = call.arguments[@"model"];
    MLKFirebaseModelSource *firebaseModelSource = [[MLKFirebaseModelSource alloc] initWithName:modelTag];
    MLKCustomRemoteModel *model = [[MLKCustomRemoteModel alloc] initWithRemoteModelSource:firebaseModelSource];
    genericModelManager = [[GenericModelManager alloc] init];
    [genericModelManager manageModel:model call:call result:result];
}
#endif

@end

File: packages/google_mlkit_object_detection/ios/google_mlkit_object_detection/Classes/GoogleMlKitObjectDetectionPlugin.swift
import Flutter
import MLKitCommon
import MLKitVision
import MLKitObjectDetection
import MLKitObjectDetectionCommon
import MLKitObjectDetectionCustom
import google_mlkit_commons

#if canImport(MLKitLinkFirebase)
import MLKitLinkFirebase
#endif

@objc
public class GoogleMlKitObjectDetectionPlugin: NSObject, FlutterPlugin {
  private var instances: [String: ObjectDetector] = [:]
  #if canImport(MLKitLinkFirebase)
  private var genericModelManager: GenericModelManager?
  #endif

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "google_mlkit_object_detector",
      binaryMessenger: registrar.messenger()
    )
    let instance = GoogleMlKitObjectDetectionPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "vision#startObjectDetector":
      handleDetection(call: call, result: result)
    case "vision#closeObjectDetector":
      if let args = call.arguments as? [String: Any], let uid = args["id"] as? String {
        instances.removeValue(forKey: uid)
      }
      result(nil)
    case "vision#manageFirebaseModels":
      #if canImport(MLKitLinkFirebase)
      manageModel(call: call, result: result)
      #else
      result(FlutterError(
        code: "ERROR_MISSING_MLKIT_FIREBASE_MODELS",
        message: "Add the GoogleMLKit/LinkFirebase pod to your Podfile to use Firebase-hosted models.",
        details: nil
      ))
      #endif
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func handleDetection(call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any],
          let imageData = args["imageData"] as? [String: Any],
          let uid = args["id"] as? String else {
      result(FlutterError(code: "invalid_args", message: "Missing arguments", details: nil))
      return
    }
    guard let image = VisionImage.visionImage(from: imageData) else {
      result(FlutterError(code: "invalid_image", message: "Invalid or missing image data", details: nil))
      return
    }

    let objectDetector: ObjectDetector
    if let existing = instances[uid] {
      objectDetector = existing
    } else {
      guard let dictionary = args["options"] as? [String: Any] else {
        result(FlutterError(code: "invalid_args", message: "Missing options", details: nil))
        return
      }
      guard let detector = createDetector(from: dictionary, result: result) else { return }
      objectDetector = detector
      instances[uid] = objectDetector
    }

    objectDetector.process(image) { objects, error in
      if let error = error as NSError? {
        result(FlutterError(code: "Error \(error.code)", message: error.domain, details: error.localizedDescription))
        return
      }
      guard let objects = objects else {
        result([])
        return
      }
      let objectsData = objects.map { object -> [String: Any] in
        let labels = object.labels.map { label in
          [
            "index": label.index,
            "text": label.text,
            "confidence": label.confidence
          ] as [String: Any]
        }
        var data: [String: Any] = [
          "rect": [
            "left": Double(object.frame.origin.x),
            "top": Double(object.frame.origin.y),
            "right": Double(object.frame.origin.x + object.frame.size.width),
            "bottom": Double(object.frame.origin.y + object.frame.size.height)
          ] as [String: Double],
          "labels": labels
        ]
        if let trackingID = object.trackingID {
          data["trackingId"] = trackingID
        }
        return data
      }
      result(objectsData)
    }
  }

  private func createDetector(from dictionary: [String: Any], result: @escaping FlutterResult) -> ObjectDetector? {
    let type = dictionary["type"] as? String ?? "base"
    switch type {
    case "base":
      return ObjectDetector.objectDetector(options: getDefaultOptions(dictionary))
    case "local":
      guard let options = getLocalOptions(dictionary) else {
        result(FlutterError(code: "invalid_args", message: "Missing path for local model", details: nil))
        return nil
      }
      return ObjectDetector.objectDetector(options: options)
    case "remote":
      #if canImport(MLKitLinkFirebase)
      guard let modelName = dictionary["modelName"] as? String, !modelName.isEmpty else {
        result(FlutterError(
          code: "invalid_args",
          message: "Missing or invalid modelName for remote model",
          details: nil
        ))
        return nil
      }
      if let options = getRemoteOptions(dictionary) {
        return ObjectDetector.objectDetector(options: options)
      }
      result(FlutterError(
        code: "Error Model has not been downloaded yet",
        message: "Model has not been downloaded yet",
        details: "Model has not been downloaded yet"
      ))
      return nil
      #else
      result(FlutterError(
        code: "ERROR_MISSING_MLKIT_FIREBASE_MODELS",
        message: "Add the GoogleMLKit/LinkFirebase pod to your Podfile to use Firebase-hosted models.",
        details: nil
      ))
      return nil
      #endif
    default:
      result(FlutterError(
        code: type,
        message: "Invalid model type: \(type)",
        details: "Invalid model type: \(type)"
      ))
      return nil
    }
  }

  private func getDefaultOptions(_ dictionary: [String: Any]) -> ObjectDetectorOptions {
    let options = ObjectDetectorOptions()
    let mode = (dictionary["mode"] as? NSNumber)?.intValue ?? 0
    options.detectorMode = mode == 0 ? ObjectDetectorMode.stream : ObjectDetectorMode.singleImage
    options.shouldEnableClassification = (dictionary["classify"] as? NSNumber)?.boolValue ?? false
    options.shouldEnableMultipleObjects = (dictionary["multiple"] as? NSNumber)?.boolValue ?? false
    return options
  }

  private func getLocalOptions(_ dictionary: [String: Any]) -> CustomObjectDetectorOptions? {
    guard let path = dictionary["path"] as? String else { return nil }
    let localModel = LocalModel(path: path)
    let options = CustomObjectDetectorOptions(localModel: localModel)
    let mode = (dictionary["mode"] as? NSNumber)?.intValue ?? 0
    options.detectorMode = mode == 0 ? ObjectDetectorMode.stream : ObjectDetectorMode.singleImage
    options.shouldEnableClassification = (dictionary["classify"] as? NSNumber)?.boolValue ?? false
    options.shouldEnableMultipleObjects = (dictionary["multiple"] as? NSNumber)?.boolValue ?? false
    if let threshold = dictionary["threshold"] as? NSNumber {
      options.classificationConfidenceThreshold = threshold
    }
    if let maxLabels = dictionary["maxLabels"] as? NSNumber {
      options.maxPerObjectLabelCount = maxLabels.intValue
    }
    return options
  }

  #if canImport(MLKitLinkFirebase)
  private func getRemoteOptions(_ dictionary: [String: Any]) -> CustomObjectDetectorOptions? {
    guard let modelName = dictionary["modelName"] as? String else { return nil }
    let firebaseModelSource = FirebaseModelSource(name: modelName)
    let remoteModel = CustomRemoteModel(remoteModelSource: firebaseModelSource)
    guard ModelManager.modelManager().isModelDownloaded(remoteModel) else {
      return nil
    }
    let options = CustomObjectDetectorOptions(remoteModel: remoteModel)
    let mode = (dictionary["mode"] as? NSNumber)?.intValue ?? 0
    options.detectorMode = mode == 0 ? ObjectDetectorMode.stream : ObjectDetectorMode.singleImage
    options.shouldEnableClassification = (dictionary["classify"] as? NSNumber)?.boolValue ?? false
    options.shouldEnableMultipleObjects = (dictionary["multiple"] as? NSNumber)?.boolValue ?? false
    if let threshold = dictionary["threshold"] as? NSNumber {
      options.classificationConfidenceThreshold = threshold
    }
    if let maxLabels = dictionary["maxLabels"] as? NSNumber {
      options.maxPerObjectLabelCount = maxLabels.intValue
    }
    return options
  }

  private func manageModel(call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any],
          let modelTag = args["model"] as? String else {
      result(FlutterError(code: "invalid_args", message: "Missing model argument", details: nil))
      return
    }
    let firebaseModelSource = FirebaseModelSource(name: modelTag)
    let model = CustomRemoteModel(remoteModelSource: firebaseModelSource)
    let manager = GenericModelManager()
    genericModelManager = manager
    manager.manage(model: model, call: call, result: result)
  }
  #endif
}
